gateway/grpc/server.go

```diff
@@ -114,102 +56,89 @@ func NewServer(ctx context.Context, endpoint string, certs []tls.Certificate, cl
-					if cert.Subject.CommonName != cert.Issuer.CommonName {
-						return nil, fmt.Errorf("tls: invalid certificate's issuer common name: %w", err)
-					}
-
-					// 2. serial number must be in
-					if cert.SerialNumber == nil {
-						return nil, fmt.Errorf("tls: invalid certificate serial number: %w", err)
-					}
-
-					// 3. look up certificate on chain
-					var resp *ctypes.QueryCertificatesResponse
-					resp, err = cquery.Certificates(
-						ctx,
-						&ctypes.QueryCertificatesRequest{
-							Filter: ctypes.CertificateFilter{
-								Owner:  owner.String(),
-								Serial: cert.SerialNumber.String(),
-								State:  "valid",
-							},
-						},
-					)
-					if err != nil {
-						return nil, fmt.Errorf("tls: unable to fetch certificate from chain: %w", err)
-					}
-					if (len(resp.Certificates) != 1) || !resp.Certificates[0].Certificate.IsState(ctypes.CertificateValid) {
-						return nil, errors.New("tls: attempt to use non-existing or revoked certificate") // nolint: goerr113
-					}
-
-					clientCertPool := x509.NewCertPool()
-					clientCertPool.AddCert(cert)
-
-					opts := x509.VerifyOptions{
-						Roots:                     clientCertPool,
-						CurrentTime:               time.Now(),
-						KeyUsages:                 []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
-						MaxConstraintComparisions: 0,
-					}
-
-					if _, err = cert.Verify(opts); err != nil {
-						return nil, fmt.Errorf("tls: unable to verify certificate: %w", err)
-					}
+				owner, err := utils.VerifyCertChain(ctx, mtls.State.PeerCertificates, "", x509.ExtKeyUsageServerAuth, cquery)
+				if err != nil {
+					return nil, fmt.Errorf("verify cert chain: %w", err)
+				}
 
+				if owner != nil {
 					ctx = ContextWithOwner(ctx, owner)
 				}
 			}
 		}
 
-		return handler(ctx, req)
+		return h(ctx, req)
 	}
 }
 
-func (gm *grpcProviderV1) GetStatus(ctx context.Context, _ *emptypb.Empty) (*providerv1.Status, error) {
-	return gm.client.StatusV1(ctx)
-}
-
-func (gm *grpcProviderV1) StreamStatus(_ *emptypb.Empty, stream providerv1.ProviderRPC_StreamStatusServer) error {
-	bus, err := fromctx.PubSubFromCtx(gm.ctx)
-	if err != nil {
-		return err
-	}
-
-	events := bus.Sub(ptypes.PubSubTopicProviderStatus)
-
-	for {
-		select {
-		case <-gm.ctx.Done():
-			return gm.ctx.Err()
-		case <-stream.Context().Done():
-			return stream.Context().Err()
-		case evt := <-events:
-			val := evt.(providerv1.Status)
-			if err := stream.Send(&val); err != nil {
-				return err
-			}
+// TODO(andrewhare): Possibly replace this with
+// https://github.com/grpc-ecosystem/go-grpc-middleware/tree/main/interceptors/logging
+// to get full request/response logging?
+func errorLogInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req any, i *grpc.UnaryServerInfo, h grpc.UnaryHandler) (any, error) {
+		resp, err := h(ctx, req)
+		if err != nil {
+			fromctx.LogcFromCtx(ctx).Error(i.FullMethod, "err", err)
```

Comment:
requesting object like log from context within interceptor is probably not good idea from performance prospective.
lets make `errorLogInterceptor` member of `server` and instantiate log during server setup.

`h` -> `next`